### PR TITLE
Fix SVD error code handling for OpenBLAS 0.3.15+ and MKL 2022+

### LINKED ISSUE #67693: `test_svd_errors_and_warnings` tests fail on CPU
## 🐛 Bug

The tests `test_svd_errors_and_warnings_cpu_complex128`, `test_svd_errors_and_warnings_cpu_complex64`, `test_svd_errors_and_warnings_cpu_float32` and `test_svd_errors_and_warnings_cpu_float64` fail with the following traceback:

```
======================================================================
FAIL: test_svd_errors_and_warnings_cpu_complex128 (__main__.TestLinalgCPU)
----------------------------------------------------------------------
RuntimeError: falseINTERNAL ASSERT FAILED at "../aten/src/ATen/native/LinearAlgebraUtils.h":243, please report a bug to PyTorch. svd_cpu: Argument 4 has illegal value. Most certainly there is a bug in the
 implementation calling the backend library.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/casparl/eb_tmp/eb-zicdqucj/tmp5xk49djn/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 368, in instantiated_test
    result = test(self, **param_kwargs)
  File "/tmp/casparl/eb_tmp/eb-zicdqucj/tmp5xk49djn/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 769, in dep_fn
    return fn(slf, *args, **kwargs)
  File "/tmp/casparl/eb_tmp/eb-zicdqucj/tmp5xk49djn/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 769, in dep_fn
    return fn(slf, *args, **kwargs)
  File "/tmp/casparl/eb_build/PyTorch/1.10.0/foss-2021a-CUDA-11.3.1/pytorch/test/test_linalg.py", line 2903, in test_svd_errors_and_warnings
    svd(a)
AssertionError: "The algorithm failed to converge" does not match "falseINTERNAL ASSERT FAILED at "../aten/src/ATen/native/LinearAlgebraUtils.h":243, please report a bug to PyTorch. svd_cpu: Argument 4 ha
s illegal value. Most certainly there is a bug in the implementation calling the backend library."
```

## To Reproduce

1. Install PyTorch 1.10.0 from sources, building against OpenBLAS 0.3.15
2. Run `python run_test.py -i test_linalg.TestLinalgCPU -v` to run the relevant part of the test suite
3. (alternative): instead of running the test from the test suite, simply do 
```
import torch
a = torch.full((3, 3), float('nan'), dtype=torch.float32, device='cpu')
torch.svd(a)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: falseINTERNAL ASSERT FAILED at "../aten/src/ATen/native/LinearAlgebraUtils.h":243, please report a bug to PyTorch. svd_cpu: Argument 4 has illegal value. Most certainly there is a bug in the implementation calling the backend library.
```

## Expected behavior

It makes sense that torch.svd wouldn't complete on a tensor of NaN's. However, the test suite clearly expects this type of error:

```
with self.assertRaisesRegex(RuntimeError, "The algorithm failed to converge"):
    svd(a)
```
But instead, I get the error above. That error looks more serious: it seems the implementation calling the backend library makes some invalid call in this particular scenario. Note that a

```
a = torch.eye(3, 3, dtype=torch.float32, device='cpu')
torch.svd(a)
```
Runs without any issues, it's really the NaN input that breaks it.

Summarizing: one wouldn't assume a `torch.svd(a)` to complete succesfully when `a` contains NaN's, but I (and the test suite) do expect a different error.

## Environment

 - PyTorch Version (e.g., 1.0): 1.10.0
 - OS (e.g., Linux): CentOS 8
 - How you installed PyTorch (`conda`, `pip`, source): source 
 - Build command you used (if compiling from source):
```
USE_CUPTI_SO=1 PYTORCH_BUILD_VERSION=1.10.0 PYTORCH_BUILD_NUMBER=1 MAX_JOBS=72 BLAS=Eigen WITH_BLAS=open USE_FFMPEG=1 BUILD_CUSTOM_PROTOBUF=0 USE_IBVERBS=1 USE_CUDA=1 CUDNN_LIB_DIR=/tmp/sw_stack_gpu/software/cuDNN/8.2.1.32-CUDA-11.3.1/lib64 CUDNN_INCLUDE_DIR=/tmp/sw_stack_gpu/software/cuDNN/8.2.1.32-CUDA-11.3.1/include USE_SYSTEM_NCCL=1 NCCL_INCLUDE_DIR=/tmp/sw_stack_gpu/software/NCCL/2.10.3-GCCcore-10.3.0-CUDA-11.3.1/include USE_METAL=0   /tmp/sw_stack_gpu/software/Python/3.9.5-GCCcore-10.3.0/bin/python setup.py build
```
 - Python version: 3.9.5
 - CUDA/cuDNN version: CUDA: 11.3.1, cuDNN: 8.2.1.32
 - GPU models and configuration: N/A
 - Any other relevant information: I'm building against OpenBLAS 0.3.15, so that should be the CPU backend in this case.

## Additional context

I've also hit [this issue](https://github.com/pytorch/pytorch/issues/67675) due to the fact that my `numpy` was compiled against the same `OpenBLAS` version, and apparantly the behaviour of how NaN's are propagated has changed from OpenBLAS 0.3.15 onwards (see https://github.com/numpy/numpy/issues/18914). I'm not sure if the current issue is related, but since it occurs specifically when the input matrices contain NaN's, I guess it could be.

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano

### LINKED ISSUE #71645: INTERNAL ASSERT in svd_cpu
### 🐛 Describe the bug

`test_svd_errors_and_warnings` is failing in master CI:
```
RuntimeError: falseINTERNAL ASSERT FAILED at "/var/lib/jenkins/workspace/aten/src/ATen/native/LinearAlgebraUtils.h":310, please report a bug to PyTorch. svd_cpu: Argument 4 has illegal value. Most certainly there is a bug in the implementation calling the backend library.
```
https://github.com/pytorch/pytorch/runs/4898981904?check_suite_focus=true#step:17:16119

To reproduce you need to rebuild with the latest dependencies on conda with `conda upgrade --all` after which both `test_norm_extreme_values` and `test_svd_errors_and_warnings` in `test_linalg.py` fail.

### Versions

```
Collecting environment information...
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.2 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.21.3

Python version: 3.8 (64-bit runtime)
Is CUDA available: N/A
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 2060
GPU 1: NVIDIA GeForce RTX 2060

Nvidia driver version: 465.19.01
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.21.4
[pip3] torch==1.11.0a0+git5f28641
[conda] cudatoolkit               11.2.2               he111cf0_8    conda-forge
[conda] magma-cuda102             2.5.2                         1    pytorch
[conda] mkl                       2022.0.1           h8d4b97c_803    conda-forge
[conda] mkl-include               2022.0.1           h8d4b97c_803    conda-forge
[conda] numpy                     1.22.0           py38h6ae9a64_1    conda-forge
[conda] torch                     1.11.0a0+git5f28641          pypi_0    pypi
```

cc @ezyang @gchanan @zou3519 @bdhirsh @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano

### PR DESCRIPTION
**Pull request by IvanYashchuk**
Fixes https://github.com/pytorch/pytorch/issues/67693.

Reference LAPACK (used in OpenBLAS) changed info error code for svd when inputs contain non-finite numbers. In PyTorch, we raise an internal assert error for negative `info` error codes because usually, it would indicate wrong implementation. However, this is not the case with SVD now in newer versions of LAPACK. MKL (tried 2021.4.0) still gives a positive error code for this kind of input. This change aligns with the OpenBLAS and MKL behavior in our code.

**UPDATE:**
MKL 2022 has uses the latest reference LAPACK behavior and returns the same `info` as OpenBLAS 0.3.15+
This PR fixes https://github.com/pytorch/pytorch/issues/71645 that is due to the updated MKL version in CI.

### CODE REVIEW

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 1]
```diff
@@ -2997,6 +2997,18 @@ static void apply_svd(Tensor& self, Tensor& U, Tensor& S, Tensor& VT,
     // Compute S, U (optionally) and VT (optionally)
     lapackSvd<scalar_t, value_t>(jobz, m, n, self_working_ptr, lda,
                         S_working_ptr, U_working_ptr, lda, VT_working_ptr, ldvt, work_data, lwork, rwork_data, iwork_data, &info);
+
+// Reference LAPACK changed `info` behavior for inputs with non-finite values
+// Previously, it would return `info` > 0, but now it returns `info` = -4
+// Making the behaviour different from MKL
+// Here we check for the case where `info` is -4 and set it to a positive number
+// This will give the same error message as with MKL
+#if AT_BUILD_WITH_OPENBLAS()
```
Comment (Lezcano) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:3006]: Should this be unconditional?

According to the LAPACK documentation in 3.10, this is expected not only for OPENBLAS, but for all LAPACK-complaint backends. See
https://www.netlib.org/lapack/explore-html/d3/da8/group__complex16_g_esing_gaccb06ed106ce18814ad7069dcb43aa27.html

Comment (IvanYashchuk) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:3006]: PyTorch's build system doesn't allow mixing different implementations of BLAS and LAPACK. Both must be from OpenBLAS, or MKL, or Accelerate/vecLib, ... .
MKL 2021.4 is not compliant with the latest Reference LAPACK implementation.
https://www.intel.com/content/www/us/en/develop/documentation/onemkl-developer-reference-fortran/top/lapack-routines/lapack-least-squares-and-eigenvalue-problem/lapack-least-squares-eigenvalue-problem-driver/singular-value-decomposition-lapack-driver/gesdd.html?wapkw=gesdd?wapkw=gesdd
There is no documentation for Apple's Accelerate/vecLib.

There are really only two used BLAS&LAPACK implementations for CPU computing which are MKL and OpenBLAS. We can safely ignore other options until someone actually complains.

Other implementations allowed by `pytorch/cmake/Modules/FindLAPACK.cmake`:
* GotoBLAS2 - not maintained anymore
* FLAME - can't find documentation, but [the code](https://github.com/flame/libflame/blob/master/src/lapack/dec/svd/front/flamec/FLA_Svd.c) didn't change from 2014 
* AMD Core Math Library (ACML) - not maintained anymore

Switching the `-4` value of `info` for any LAPACK implementation shouldn't hurt anything.
We could also specialize SVD's error with `info = -4` to `Sorry, the input contained non-finite values and the computation cannot be completed`, but we don't have the same info from MKL to raise the same error.

**Review by lezcano — APPROVED** [review]
Fair enough then.

On a related but different topic. I still like the idea of, if we get an `info > 0`, rather than raising that obscure "wrong implementation" error, first check for `NaN`s. That would account for this case, and I suspect that also for a number of other cases.

**Comment by IvanYashchuk** [comment]
@mruberry, could you please take a look and merge if the change looks reasonable to you?

**Review by mruberry — APPROVED** [review]
Cool!

Thanks @IvanYashchuk

**Comment by mruberry** [comment]
Some internal builds are complaining with

```
stderr: aten/src/ATen/native/BatchLinearAlgebra.cpp:3006:5: error: invalid token at start of a preprocessor expression
#if AT_BUILD_WITH_OPENBLAS()
```

Not sure what's going on there

**Comment by dagitses** [comment]
> Some internal builds are complaining with
> 
> ```
> stderr: aten/src/ATen/native/BatchLinearAlgebra.cpp:3006:5: error: invalid token at start of a preprocessor expression
> #if AT_BUILD_WITH_OPENBLAS()
> ```
> 
> Not sure what's going on there

This looks like an internal build issue and some internal build files will need to define the expansion like this PR does in  BUILD.bazel.

**Comment by IvanYashchuk** [comment]
@dagitses, @mruberry is there anything on my side I could help with to merge this PR?

**Comment by mruberry** [comment]
> @dagitses, @mruberry is there anything on my side I could help with to merge this PR?

Not that I know of at this time, sorry. It's just going to require some work internally.

**Comment by IvanYashchuk** [comment]
Latest MKL fails with the same error as OpenBLAS 0.3.15: https://github.com/pytorch/pytorch/issues/71645
I need to modify this PR to cover MKL as well.

**Comment by ngimel** [comment]
@IvanYashchuk please ping me when you rebase and update PR.

File: test/test_linalg.py [hunk 2]
```diff
@@ -1910,8 +1909,8 @@ def is_broken_matrix_norm_case(ord, x):
                 if is_broken_matrix_norm_case(ord, x):
                     continue
                 else:
-                    result = torch.linalg.norm(x, ord=ord)
                     result_n = np.linalg.norm(x_n, ord=ord)
+                    result = torch.linalg.norm(x, ord=ord)
```
Comment (IvanYashchuk) [inline test/test_linalg.py:1913]: Reordered here so that whoever tries to enable skipped matrix_ords in the future first sees the failures from NumPy.

**Comment by IvanYashchuk** [comment]
@mruberry and @ngimel, I updated this PR. Now it doesn't have any changes to the build system so should be easy to merge once approved.
 As mentioned in the issue description this PR fixes https://github.com/pytorch/pytorch/issues/71645. The source of the problem is not flakiness but a recent update of CI to use the new MKL version changing the behavior of errors.

**Comment by ngimel** [comment]
@IvanYashchuk can you please rebase to fix conflict?

**Comment by IvanYashchuk** [comment]
@ngimel, the conflict is resolved.

**Comment by pbelevich** [comment]
@IvanYashchuk  reverting because:
```
======================================================================
FAIL [0.001s]: test_svd_nan_error_meta_float64 (__main__.TestLinalgMETA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 1753, in wrapper
    method(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 376, in instantiated_test
    result = test(self, **param_kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 788, in dep_fn
    return fn(slf, *args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 788, in dep_fn
    return fn(slf, *args, **kwargs)
  File "test_linalg.py", line 2628, in test_svd_nan_error
    svd(a)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 1639, in __exit__
    return super().__exit__(exc_type, exc_value, tb)
AssertionError: _LinAlgError not raised
```

File: test/test_linalg.py [hunk 3]
```diff
@@ -2614,6 +2613,26 @@ def test_svd(self, device, dtype):
...
+            error_msg = r'(CUSOLVER_STATUS_EXECUTION_FAILED|The algorithm failed to converge)'
+            a = torch.full((3, 3), float('nan'), dtype=dtype, device=device)
+            a[0] = float('nan')
+            with self.assertRaisesRegex(torch.linalg.LinAlgError, error_msg):
+                svd(a)
+            error_msg = r'(CUSOLVER_STATUS_EXECUTION_FAILED|\(Batch element 1\): The algorithm failed to converge)'
+            a = torch.randn(3, 33, 33, dtype=dtype, device=device)
+            a[1, 0, 0] = float('nan')
+            with self.assertRaisesRegex(torch.linalg.LinAlgError, error_msg):
+                svd(a)
```
Comment (Lezcano) [inline test/test_linalg.py:2634]: Thank you for saving this part and catching my mistake!

**Review by lezcano — APPROVED** [review]

**Comment by jeffdaily** [comment]
This PR was merged, reverted, and now merged again.  The first merge attempt broke cuda but that seems fixed with the latest merge.  But rocm CI was broken by the first merge as well as the latest merge.  Do we revert again and fix rocm CI?

**Comment by IvanYashchuk** [comment]
Alright, ROCm build is now successful: https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm4.5-py3.7-trigger/5021/ and all other CI jobs are green as well.

**Comment by IvanYashchuk** [comment]
@mruberry, I modified this PR slightly so that there are no conflicts with https://github.com/pytorch/pytorch/pull/72125.

**Comment by IvanYashchuk** [comment]
This PR fixes false triggers of `TORCH_INTERNAL_ASSERT` that suggest users file a bug report (see https://github.com/pytorch/pytorch/issues/71645). Therefore I think it's important to include this PR in the next 1.11 release, users should see a nicer error message.

**Comment by IvanYashchuk** [comment]
@ngimel, could you please import this PR?  Here is [the HUD for the CI](https://hud.pytorch.org/pr/pytorch/pytorch/68812), failed tests on backwards compatibility are not related.

**Comment by ngimel** [comment]
Can you please open a new PR? I can't reimport this one.

**Pull request closed without merging**